pkg/providers/amifamily/ami.go

```diff
@@ -54,18 +55,37 @@ type DefaultProvider struct {
 }
 
 type AMI struct {
-	Name         string
-	AmiID        string
-	CreationDate string
-	Requirements scheduling.Requirements
+	Name            string
+	AmiID           string
+	CreationDate    string
+	DeprecationTime string
+	Requirements    scheduling.Requirements
 }
 
 type AMIs []AMI
 
 // Sort orders the AMIs by creation date in descending order.
-// If creation date is nil or two AMIs have the same creation date, the AMIs will be sorted by ID, which is guaranteed to be unique, in ascending order.
+// If there are any amis with deprecation time, they will be sorted by deprecation time in descending order
+// If there deprecation time is nil or two AMIs have the same deprecation time, the AMIs will be sorted by ID, which is guaranteed to be unique, in ascending order.
+// finally, if creation date is nil or two AMIs have the same creation date, the AMIs will be sorted by ID, which is guaranteed to be unique, in ascending order.
 func (a AMIs) Sort() {
+	maxDepTime := time.Unix(math.MaxInt64, 0)
 	sort.Slice(a, func(i, j int) bool {
+		ideptime := lo.TernaryF(
+			a[i].DeprecationTime != "",
+			func() time.Time { t, _ := time.Parse(time.RFC3339, a[i].DeprecationTime); return t },
```

Comment:
Fail loudly if this invariant is violated
```suggestion
			func() time.Time { t := lo.Must(time.Parse(time.RFC3339, a[i].DeprecationTime)); return t },
```

Reply:
Thank you! Updated :)

Reply:
@ellistarn how is this different from the existing, where errors are ignored? Should that as well become `lo.Must`?
https://github.com/aws/karpenter-provider-aws/blob/8bc203024b3c45739d3e928bbf71365530843375/pkg/providers/amifamily/ami.go#L89-L90

Reply:
question - this is the same for the existing piece for `CreationDate` where we have the following lines 

```
		itime, _ := time.Parse(time.RFC3339, a[i].CreationDate)
		jtime, _ := time.Parse(time.RFC3339, a[j].CreationDate)
		if itime.Unix() != jtime.Unix() {
			return itime.Unix() > jtime.Unix()
		}
		return a[i].AmiID < a[j].AmiID
```